compiler/passes/errorHandling.cpp

```diff
@@ -546,30 +668,48 @@ bool CanThrowVisitor::enterCallExpr(CallExpr* node) {
 
 } /* end anon namespace */
 
-
 // Returns `true` if a block can exit with an error
 //  (e.g. by calling 'throw' or a throwing function,
 //   when these are not handled by try! or catch).
 // This function is useful to infer 'throws' for
 // certain compiler-introduced functions.
 
-static bool canBlockThrow(BlockStmt* block)
+static void markImplicitThrows(FnSymbol* fn, std::set<FnSymbol*>* visited, implicitThrowsReasons_t* reasons)
 {
-  CanThrowVisitor visit(false, false);
+  // Currently, only task functions or iterators can be implicitly throws.
+  if (!isTaskFun(fn) && !fn->isIterator())
+    return;
+
+  // If we already visited this function, don't visit it again.
+  if (visited->count(fn) > 0)
+    return;
 
-  block->accept(&visit);
+  // Add to visited set first thing to prevent infinite recursion
+  // if there ever is a cycle of calls.
+  visited->insert(fn);
 
-  return visit.throws();
+  CanThrowVisitor visit(visited, reasons);
+
+  // Note that this function can recurse because
+  // the CanThrowVisitor will call markImplicitThrows.
+  fn->body->accept(&visit);
+
+  if (visit.throws()) {
+    (*reasons)[fn] = visit.reason();
+    fn->throwsErrorInit();
+
+    if (visit.unchecked())
+      fn->addFlag(FLAG_UNCHECKED_THROWS);
+  }
 }
 
-static void checkErrorHandling(FnSymbol* fn)
+static void checkErrorHandling(FnSymbol* fn, implicitThrowsReasons_t* reasons)
```

Comment:
Same as `checkImplicitThrows`, maybe `checkErrorHandling` makes more sense as a static method of `ErrorCheckingVisitor`.